custom_components/unifiprotect/config_flow.py

```diff
@@ -40,51 +41,93 @@ def async_get_options_flow(config_entry):
         session = async_create_clientsession(
             self.hass, cookie_jar=CookieJar(unsafe=True)
         )
 
-        unifiprotect = UpvServer(
-            session,
-            user_input[CONF_HOST],
-            user_input[CONF_PORT],
-            user_input[CONF_USERNAME],
-            user_input[CONF_PASSWORD],
+        if self.unique_id is not None:
+            entry = await self.async_set_unique_id(self.unique_id)
+            host = entry.data[CONF_HOST]
+            port = entry.data[CONF_PORT]
+        elif CONF_HOST in user_input and CONF_PORT in user_input:
+            host = user_input[CONF_HOST]
+            port = user_input[CONF_PORT]
+        else:
+            return None
+
+        return UpvServer(
+            session=session,
+            host=host,
+            port=port,
+            username=user_input[CONF_USERNAME],
+            password=user_input[CONF_PASSWORD],
         )
 
+    async def _reload_entry(self, unique_id, user_input):
+        entry = await self.async_set_unique_id(unique_id)
+
+        new_data = entry.data.copy()
+        new_data[CONF_USERNAME] = user_input[CONF_USERNAME]
+        new_data[CONF_PASSWORD] = user_input[CONF_PASSWORD]
+
+        self.hass.config_entries.async_update_entry(entry, data=new_data)
+        await self.hass.config_entries.async_reload(entry.entry_id)
+        return self.async_abort(reason="reauth_successful")
+
+    async def _init_config_entry(self, user_input=None, reauth=False):
+        """Common method to initalize config entry"""
+
+        form = self._get_form(reauth)
+        if user_input is None:
+            return await form(user_input)
+
+        errors = {}
+        protect = await self._get_protect(user_input)
+
+        if protect is None:
+            return await form(errors)
+
         try:
-            server_info = await unifiprotect.server_information()
+            server_info = await protect.server_information()
             if server_info["server_version"] < MIN_REQUIRED_PROTECT_V:
                 _LOGGER.debug("UniFi Protect Version not supported")
                 errors["base"] = "protect_version"
-                return await self._show_setup_form(errors)
+                return await form(errors)
 
         except NotAuthorized as ex:
             _LOGGER.debug(ex)
             errors["base"] = "connection_error"
-            return await self._show_setup_form(errors)
+            return await form(errors)
         except NvrError as ex:
             _LOGGER.debug(ex)
             errors["base"] = "nvr_error"
-            return await self._show_setup_form(errors)
+            return await form(errors)
 
         unique_id = server_info[SERVER_ID]
-        server_name = server_info[SERVER_NAME]
+        if reauth:
+            return await self._reload_entry(unique_id, user_input)
 
-        await self.async_set_unique_id(unique_id)
         self._abort_if_unique_id_configured()
```

Comment:
The branching is a bit hard to follow above.

Its not immediately clear where unique id is being set before this is called

Reply:
I have refactored out some core bits so I am going to back through and split this back into two functions based on your you recommendation about the fall through.